OSDOCS-10856: Add bpfman Operator (#80208)
openshift/openshift-docs: networking/network_security/bpfman-operator/bpfman-operator-about.adoc

:_mod-docs-content-type: ASSEMBLY
[id="bpfman-operator-about"]
= About the bfpman Operator
include::_attributes/common-attributes.adoc[]
:context: bpfman-operator-about

toc::[]

:FeatureName: bpfman Operator
include::snippets/technology-preview.adoc[]

> == About Extended Berkeley Packet Filter (eBPF)

Reviewer: [needs an id](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-titles-and-section-headings). maybe something like:
```suggestion
[id="about-ebpf_{context}"]
== About Extended Berkeley Packet Filter (eBPF)
```

eBPF extends the original Berkeley Packet Filter for advanced network traffic filtering. It acts as a virtual machine inside the Linux kernel, allowing you to run sandboxed programs in response to events such as network packets, system calls, or kernel functions.

> == About the bpfman Operator

Reviewer: ```suggestion
[id="about-bpfman-operator_{context}"]
== About the bpfman Operator
```

eBPF CRDs: A set of CRDs like XdpProgram and TcProgram for loading eBPF programs, and a bpfman-generated CRD (BpfProgram) for representing the state of loaded programs.
bpfman-agent:: Runs within a daemonset container, ensuring eBPF programs on each node are in the desired state.
bpfman-operator:: Manages the lifecycle of the bpfman-agent and CRDs in the cluster using the Operator SDK.

The bpfman Operator offers the following features:

- Enhances security by centralizing eBPF program loading through a controlled daemon. Access is regulated by standard Kubernetes RBAC, and rules for the ordering of networking eBPF programs can be defined.
- Provides detailed visibility into active eBPF programs, improving debuggability across the system.
- Facilitates the coexistence of multiple eBPF programs from different sources using protocols like libxdp for XDP and TC programs, enhancing interoperability.
- Streamlines the deployment and lifecycle management of eBPF programs in Kubernetes. Developers can focus on program interaction rather than lifecycle management, with support for existing eBPF libraries like Cilium, libbpf, and Aya.

> [id="{context}-additional-resources"]

Reviewer: [id format](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies) and role:
```suggestion
[role="_additional-resources"]
[id="additional-resources_{context}"]
```

== Additional resources

* link:https://ebpf.io/what-is-ebpf/[eBPF Documentation]
* link:https://bpfman.io/latest[bpfman]
* link:https://bpfman.io/v0.5.0/developer-guide/api-spec/[bpfman custom resource definition (CRD) API specification]

> [id="{context}-next-steps"]

Reviewer: ```suggestion
[id="next-steps_{context}"]
```

== About the bpfman Operator

The open source project bpfman simplifies the security, management, and deployment of eBPF programs within Kubernetes. It utilizes Kubernetes custom resource definitions (CRDs) to manage eBPF programs packaged as OCI container images. This approach helps to delineate deployment permissions and enhance security by restricting program types deployable by specific users.

bpfman is a software stack designed to manage eBPF programs within Kubernetes. It facilitates the loading, unloading, modifying, and monitoring of eBPF programs in Kubernetes clusters. It includes a daemon, CRDs, an agent, and an operator:

bpfman:: A system daemon that interacts with eBPF programs via a gRPC API.
eBPF CRDs: A set of CRDs like XdpProgram and TcProgram for loading eBPF programs, and a bpfman-generated CRD (BpfProgram) for representing the state of loaded programs.
bpfman-agent:: Runs within a daemonset container, ensuring eBPF programs on each node are in the desired state.
bpfman-operator:: Manages the lifecycle of the bpfman-agent and CRDs in the cluster using the Operator SDK.

> 

Reviewer 2: This whole section I think needs to be either reworked or removed.  We shouldn't call this an "open source project".  I assume this was taken from the bpfman.io site but this is not appropriate for our downstream/product documentation.

Author: @davegord, sure, I'm open to suggestions for what content to include here. Thanks!

Reviewer 3: I think it's based on some upstream content, but not word for word.  Is anyone going to take a pass at it?

Author: @anfredette, a pass, in so far as, this content is incorrect or not relevant here? Thanks!

[id="about-bpfman-operator_{context}"]
== About the bpfman Operator

The open source project bpfman simplifies the security, management, and deployment of eBPF programs within Kubernetes. It utilizes Kubernetes custom resource definitions (CRDs) to manage eBPF programs packaged as OCI container images. This approach helps to delineate deployment permissions and enhance security by restricting program types deployable by specific users.

bpfman is a software stack designed to manage eBPF programs within Kubernetes. It facilitates the loading, unloading, modifying, and monitoring of eBPF programs in Kubernetes clusters. It includes a daemon, CRDs, an agent, and an operator:

bpfman:: A system daemon that interacts with eBPF programs via a gRPC API.
eBPF CRDs: A set of CRDs like XdpProgram and TcProgram for loading eBPF programs, and a bpfman-generated CRD (BpfProgram) for representing the state of loaded programs.
bpfman-agent:: Runs within a daemonset container, ensuring eBPF programs on each node are in the desired state.
bpfman-operator:: Manages the lifecycle of the bpfman-agent and CRDs in the cluster using the Operator SDK.

> The bpfman Operator offers the following features:

Reviewer 4: @dave-tucker This is where we list the advantages/features of bpfman. Can you help me refine this list?

> - Enhances security by centralizing eBPF program loading through a controlled daemon. Access is regulated by standard Kubernetes RBAC, and rules for the ordering of networking eBPF programs can be defined.

Reviewer 4: I think we want to specifically call out that by using bpfman to load your eBPF programs, application pods don't need to be privileged. Don't necessarily take what I wrote here word for word, just giving an idea on what might need to be here.

```suggestion
- Enhances security by centralizing eBPF program loading through a controlled daemon. bpfman has the elevated privileges so the applications don't need to be. eBPF program control is regulated by standard Kubernetes RBAC, which can allow or deny applications access to the different bpfman CRDs which manage eBPF program loading and unloading.
```
